chainstate/src/detail/transaction_verifier/mod.rs

```diff
@@ -256,66 +282,71 @@ impl<'a, S: BlockchainStorageRead> TransactionVerifier<'a, S> {
                         })?;
 
-                    Self::get_output_value(tx.outputs(), output_index, tx.get_id().into())
-                        .cloned()?
+                    let output_value =
+                        Self::get_output_value(tx.outputs(), output_index, tx.get_id().into())?;
+
+                    let (key, amount) = filter_for_total_inputs(output_value, &tx)?;
+                    insert_or_increase(&mut result, key, amount)?
                 }
                 common::chain::SpendablePosition::BlockReward(block_id) => {
                     let outputs = self.block_reward_outputs(block_id)?;
-                    Self::get_output_value(&outputs, output_index, (*block_id).into()).cloned()?
-                }
-            };
-            match output_amount {
-                OutputValue::Coin(output_amount) => {
-                    total = (total + output_amount)
-                        .ok_or(ConnectTransactionError::InputAdditionError)?
+                    match Self::get_output_value(&outputs, output_index, (*block_id).into())? {
+                        OutputValue::Coin(amount) => {
+                            insert_or_increase(&mut result, CoinOrTokenId::Coin, *amount)?
+                        }
+                        OutputValue::Token(_) => {
+                            return Err(ConnectTransactionError::TokensError(
+                                TokensError::BlockRewardInTokens,
+                            ));
+                        }
+                    }
                 }
             }
         }
-        Ok(total)
+        Ok(result)
+    }
+
+    fn get_paid_fee_unchecked(
+        inputs_total_map: &BTreeMap<CoinOrTokenId, Amount>,
+        outputs_total_map: &BTreeMap<CoinOrTokenId, Amount>,
+    ) -> Result<Amount, ConnectTransactionError> {
+        let outputs_total =
+            *outputs_total_map.get(&CoinOrTokenId::Coin).unwrap_or(&Amount::from_atoms(0));
+        let inputs_total =
+            *inputs_total_map.get(&CoinOrTokenId::Coin).unwrap_or(&Amount::from_atoms(0));
+        (inputs_total - outputs_total).ok_or(ConnectTransactionError::TxFeeTotalCalcFailed(
+            inputs_total,
+            outputs_total,
+        ))
     }
 
     fn check_transferred_amounts_and_get_fee(
         &self,
+        block_id: Id<Block>,
         tx: &Transaction,
     ) -> Result<Amount, ConnectTransactionError> {
-        let inputs = tx.inputs();
-        let outputs = tx.outputs();
+        let inputs_total_map = self.calculate_total_inputs(tx.inputs())?;
+        let outputs_total_map = Self::calculate_total_outputs(tx.outputs())?;
 
-        let inputs_total = self.calculate_coins_total_inputs(inputs)?;
-        let outputs_total = Self::calculate_coins_total_outputs(outputs)?;
+        let total_fee = check_transferred_amount(&inputs_total_map, &outputs_total_map)
+            .and_then(|_| Self::get_paid_fee_unchecked(&inputs_total_map, &outputs_total_map))?;
```

Comment:
I don't understand. Why not use a question mark on the check_transferred_amount and then call this independently? Why bind them together like this?

Reply:
That was my extra paranoia, I have a look again here and yes we can use just 

```Rust
        check_transferred_amount(&inputs_total_map, &outputs_total_map)?;
        let total_fee = Self::get_paid_fee_unchecked(&inputs_total_map, &outputs_total_map)?;
```